README.txt
$HeadURL: https://xtpl.svn.sourceforge.net/svnroot/xtpl/trunk/README.txt $
$Id: README.txt 16 2007-01-11 03:02:49Z cocomp $

		Welcome to XTemplate: a templating engine for PHP.
				http://www.phpxtemplate.org/

Purpose:

	- To enable the abstraction of PHP code logic from the display layer (HTML
	  etc).
	Or:
	- To enable coders to do what they do best, code and designers to
	  concentrate on their forte, presentation.

Features:

	- Simple variable interpolation ( {STUFF} )

	- Array variable interpolation
	 	e.g. {ROW.ID} {ROW.NAME} {ROW.WHATEVER}, you assign an array to the
	 	variable ROW once

	- Global PHP variables available without making assigns
	 	e.g. {PHP._SERVER.HTTP_HOST} {PHP._SERVER.PHP_SELF} or any variable:
	 	{PHP._COOKIE.PHPSESSID}

	- Assign using arrays
		e.g. $xtpl->assign(array('STUFF' => 'kewlstuff', 'ROW' => $row));

	- Default nullstring for unassigned variables and unparsed blocks,
	 	custom nullstring for every variable or block

	- Dynamic blocks:
		- Multiple level dynamic blocks (nested blocks)
		- Autoreset feature (do not reset subblocks when reparsing parent
		  block)
		- Recursive parse to parse all subblocks of a block

	- File (template) include: {FILE "another.template.xtpl"}, the included
		- templates can also contain includes recursively..
		- uses fopen so you can include http:// and ftp:// files too

	- {FILE {VAR}} - you can assign filenames on the fly for {VAR}

	- On big projects you have the ability to substitute files with some other
	  templates

	- Very simple and powerful syntax, you'll love it :)

Documentation:

	For first time documentation, see the included example files (ex1.php -
	ex7.php) and their associated templates (*.xtpl)
		(cranx tried to write them so even newbies should understand easily)

	You should also find a docs folder with this distribution containing
	phpDocumentor (http://www.phpdoc.org) output.

	Other documenation can be found at http://www.phpxtemplate.org/

Installation:

	Nice and straight forward, just have upload it and include it - that's it!

Upgrade:

	(stating the obvious - but some people seem to forget!)

	As with any software evaluate the upgraded version on some test code before
	deployment, then backup the previous version - in case there's any
	problems, then copy the new version over the old. Test, rollback to your
	backup (you did make a backup didn't you?!?) version if there's an issue,
	check the bug list and report if it's a new bug you've discovered.

History:

The code is very short and very optimized, according to my speed tests it's the
fastest template engine around which can do nested blocks and variables..

The basic syntax is from FastTemplate and QuickTemplate
(http://phpclasses.upperdesign.com/browse.html/package/49), but the entire
class was written by cranx from scratch in one day, without a line from other
template engines. THIS IS NOT A REWRITE OF OTHER ENGINES!

The algorithm used makes this code amazing fast, you can do a bunch of
nested dynamic blocks and everything you want, because it doesn't use recursive
calls to build up the block tree.

Docs and some functions we didn't need (clearing variables, etc) are still
missing, but they'll come if there's demand.

Sometime around 2002, cranx was too busy to continue with the project, so
cocomp took over (kind of!), cocomp was then too busy for about, erm, 3 years!
Anyway, cocomp has found a new lease of life for the project and rolled up the
last few years improvements and released them on Sourceforge. Enjoy. If you
want to get involved in XTemplate, see http://www.phpxtemplate.org/

Latest stable & Subversion versions always available @
http://sourceforge.net/projects/xtpl/

Copyright (c) 2000-2001 Barnabas Debreceni [cranx@users.sourceforge.net]
Copyright (c) 2002-2007 Jeremy Coates [cocomp@users.sourceforge.net]

Licensed BSD / LGPL - see license.txt

--
END

changelog.txt
/**
 * $HeadURL: https://xtpl.svn.sourceforge.net/svnroot/xtpl/trunk/changelog.txt $
 * $Id: changelog.txt 21 2007-05-29 18:01:15Z cocomp $
 *
 * Changelog for XTemplate & CachingXTemplate:
 * For latest log information always use svn log <filename> in a checked out working copy
 * or svn log https://xtpl.svn.sourceforge.net/svnroot/xtpl/trunk/<filename> for direct access
 *
 * r21 | cocomp | 2007-05-29 19:00:00 +0000 (Tue, 29 May 2007) | 1 line
 * Added facility for constructor $tpldir to be an array of file paths, the first path supplied that has the file is matched
 *
 * r20 | cocomp | 2007-01-11 03:39:26 +0000 (Thu, 11 Jan 2007) | 1 line
 * Initial commit CachingXTemplate subclass
 *
 * r18 | cocomp | 2007-01-11 03:29:11 +0000 (Thu, 11 Jan 2007) | 2 lines
 * Major update for initial PHP 5 version
 * Deprecated: PHP 4 constructor, SetNullString (use set_null_string) and SetNullBlock (use set_null_block). Added sf Feature request #1529478. Fixed SF Bug #1261828. Introduced comment_preg property. Refactored assign method - may break some old functionality! Constructor now only calls restart method
 *
 * r12 | cocomp | 2006-05-02 15:47:11 +0100 (Tue, 02 May 2006) | 2 lines
 * Removed new line at end of file
 *
 * r11 | cocomp | 2006-05-02 15:27:19 +0100 (Tue, 02 May 2006) | 7 lines
 * 2006/05/02:
 *
 * - Added $over_write parameter to assign method which defaults to false. The
 *  behaviour is therefore to always overwrite the name/value array
 * - Fixed bug in regexes for older php version (specifically php 4.2.2) that
 *  caused regex to fail
 *
 * r10 | cocomp | 2005-08-17 10:57:58 +0100 (Wed, 17 Aug 2005) | 11 lines
 * Changelog
 * 2005/08/17:
 *
 * - Improved phpDoc entries
 * - Added support for XTPL_DIR constant from SF Feature request 1202027:  Kenneth Kalmer
 * - Added support for comments in {FILE } includes e.g. {FILE "somefile.xtpl"# An included file}
 * - Rewritten assign method, now checks if $val is an array instead of $name which never would be. Also added new parameter for stripslashes support e.g. $xtpl->assign('var', 'var', get_magic_quotes_qpc());
 * - Moved most regular expressions to use | as delimiter
 * - Improved output with regard to block tags and line endings - much better for xml output
 * - Added support for empty files in file includes - previously required at least one byte
 *
 * r9 | cocomp | 2005-04-11 12:20:28 +0100 (Mon, 11 Apr 2005) | 2 lines
 * Fixed backslashes issue (properly\!)
 *
 * r8 | cocomp | 2005-04-11 11:00:47 +0100 (Mon, 11 Apr 2005) | 2 lines
 * Added restart() method sf:641407 feature request
 *
 * r7 | cocomp | 2005-04-08 10:17:37 +0100 (Fri, 08 Apr 2005) | 2 lines
 * Fixed bug with backslashes sf:810773 & updated docs
 *
 * r5 | cocomp | 2005-04-07 13:02:52 +0100 (Thu, 07 Apr 2005) | 2 lines
 * MAJOR UPDATE: E_ALL safe, better internal documentation, code readability ++, many bugfixes and new features - considered stable
 *
 * r4 | cocomp | 2005-04-06 14:22:44 +0100 (Wed, 06 Apr 2005) | 2 lines
 * Renamed xtemplate.inc to xtemplate.class.php
 */

/**
 * Contributors:
 *
 * Original Author / Maintainer:
 * Barnabas Debreceni [cranx@users.sourceforge.net] Copyright (c) 2000-2001
 *
 * Current Author / Maintainer:
 * Jeremy Coates [cocomp@users.sourceforge.net] Copyright (c) 2002-2007
 *
 * Ivar Smolin <okul@linux.ee> (14-march-2001)
 * - made some code optimizations
 * Bert Jandehoop <bert.jandehoop@users.info.wau.nl> (26-june-2001)
 * - new feature to substitute template files by other templates
 * - new method array_loop()
 *
 * Various contributions over the years from:
 * Code: Noel Walsh (NW), John Carter (JC)
 * Bug reporting: SadGeezer
 */

/**
 * Changelog Archive
 *
 *****
 *
 * Stuff from development outside sourceforge
 *
 * Revision 1.2  2003/12/05 22:22:17  jeremy
 * Removed duplicate function call in out method
 *
 * Revision 1.1.1.1  2003/10/29 20:22:43  jeremy
 * Initial Import
 *
 * Revision 1.1  2003/06/25 17:17:52  jeremy
 * Initial Import
 *
 * Revision 1.4  2001/08/17 18:25:45  jeremy
 * Sorted greedy matching regular expression in parse function preg_match_all line 166: added ? after .* when looking for comments
 *
 *****
 *
 * Old log stuff
 *
 * Revision 1.2  2001/09/19 14:11:25  cranx
 * fixed a bug in the whitespace-stripping block variable interpolating regexp.
 *
 * Revision 1.1  2001/07/11 10:42:39  cranx
 * added:
 * - filename substitution, no nested arrays for the moment, sorry
 * (including happens when assigning, so assign filevar in the outside blocks first!)
 *
 * Revision 1.5  2001/07/11 10:39:08  cranx
 * added:
 * - we can now specify base dir
 * - array_loop()
 * - trigger_error in _set_error
 *
 * modified:
 * - newline bugs fixed (for XML)
 * - in out(): content-length header added
 * - whiles changed to foreach
 * - from now on, the class is php4 only :P
 *
 *****
 *
 * Old stuff from original releases
 *
 * xtemplate class 0.3pre
 * !!! {FILE {VAR}} file variable interpolation may still be buggy !!!
 */

caching_xtemplate.class.php
<?php

require_once(dirname(__FILE__) . DIRECTORY_SEPARATOR . 'xtemplate.class.php');

/**
 * CachingXTemplate
 * Extension to XTemplate to provide block level and whole template caching facilities
 * Needs Web server writable directory
 *
 * @package XTemplate
 * @subpackage CachingXTemplate
 * @uses XTemplate
 * @author Jeremy Coates [cocomp@users.sourceforge.net]
 * @copyright Jeremy Coates / Co-Comp Ltd 2006-2007
 * @see license.txt BSD license
 * @since PHP 5
 * @link $HeadURL: https://xtpl.svn.sourceforge.net/svnroot/xtpl/trunk/caching_xtemplate.class.php $
 * @version $Id: caching_xtemplate.class.php 21 2007-05-29 18:01:15Z cocomp $
 *
 * @example Whole template level caching (e.g. the total parsed output for the file)
 * @example $xtpl = new CachingXTemplate('template.xtpl', '', null, 'main', true, 600, session_id(), './xcache', '.xcache');
 *
 * @example Alternatively (and perhaps more useful in real world):
 * @example Block level caching
 * @example $xtpl = new CachingXTemplate('template.xtpl', '', null, 'main', true, 0, session_id(), './xcache', '.xcache');
 * @example $xtpl->parse('main', 600);
 * @example Bear in mind that because XTemplate uses a reversed parsing tree the innermost blocks need to be parsed
 * @example first, therefore if you cache an outer block, don't be surprised when it's inner content blocks don't update!
 */
class CachingXTemplate extends XTemplate {

	/**
	 * Cache expiry time (seconds)
	 *
	 * @access public
	 * @var int
	 */
	public $cache_expiry	= 0;

	/**
	 * Cache file unique identifier
	 *
	 * @example session_id()
	 * @access public
	 * @var string
	 */
	public $cache_unique	= 'unique';

	/**
	 * Filename extension
	 *
	 * @example .xcache
	 * @access public
	 * @var string
	 */
	public $cache_ext		= '.xcache';

	/**
	 * Path to cache dir
	 * Needs to be writable by webserver
	 *
	 * @example ./xcache
	 * @access public
	 * @var string
	 */
	public $cache_dir		= './xcache';

	/**
	 * Flag showing whether template is cached
	 *
	 * @access private
	 * @var boolean
	 */
	private $_template_is_cached	= false;

	/**
	 * Cache expiry time
	 *
	 * @access private
	 * @var int
	 */
	private $_cache_expiry			= 0;

	/**
	 * File modified time
	 *
	 * @access private
	 * @var int
	 */
	private $_cache_filemtime		= 0;

	/**
	 * Override of parent constructor
	 *
	 * @access public
     * @param string $file Template file to work on
     * @param string $tpldir Location of template files (useful for keeping files outside web server root)
     * @param array $files Filenames lookup
     * @param string $mainblock Name of main block in the template
     * @param boolean $autosetup If true, run setup() as part of constuctor
	 * @param int $cache_expiry Seconds to cache for
	 * @param string $cache_unique Unique file id (e.g. session_id())
	 * @param string $cache_dir Cache folder
	 * @param string $cache_ext Cache file extension
	 */
	public function __construct($file, $tpldir = '', $files = null, $mainblock = 'main', $autosetup = true, $cache_expiry = 0, $cache_unique = '', $cache_dir = './xcache', $cache_ext = '.xcache') {

		$this->restart($file, $tpldir, $files, $mainblock, $autosetup, $this->tag_start_delim, $this->tag_end_delim, $cache_expiry, $cache_unique, $cache_dir, $cache_ext);

	}

	/**
	 * Override of parent restart method
	 *
	 * @access public
	 * @param string $file Template file to work on
	 * @param string $tpldir Location of template files
	 * @param array $files Filenames lookup
	 * @param string $mainblock Name of main block in the template
	 * @param boolean $autosetup If true, run setup() as part of restarting
	 * @param string $tag_start {
	 * @param string $tag_end }
	 * @param int $cache_expiry Seconds to cache for
	 * @param string $cache_unique Unique file id (e.g. session_id())
	 * @param string $cache_dir Cache folder
	 * @param string $cache_ext Cache file extension
	 */
	public function restart ($file, $tpldir = '', $files = null, $mainblock = 'main', $autosetup = true, $tag_start = '{', $tag_end = '}', $cache_expiry = 0, $cache_unique = '', $cache_dir = './xcache', $cache_ext = '.xcache') {

		if ($cache_expiry > 0) {
			$this->cache_expiry = $cache_expiry;
		}

		if (!empty($cache_unique)) {
			if (!preg_match('/^\./', $cache_unique)) {
				$cache_unique = '.' . $cache_unique;
			}
			$this->cache_unique = $cache_unique;
		}

		if (!empty($cache_dir)) {
			$this->cache_dir = $cache_dir;
		}

		if (!empty($cache_ext)) {
			if (!preg_match('/^\./', $cache_ext)) {
				$cache_ext = '.' . $cache_ext;
			}
			$this->cache_ext = $cache_ext;
		}

		// Call parent restart method but don't run setup yet!
		parent::restart($file, $tpldir, $files, $mainblock, false, $tag_start, $tag_end);

		if ($this->cache_expiry > 0) {
			$this->read_template_cache();
		}

		if (!$this->_template_is_cached && $autosetup) {
			$this->setup();
		}
	}

	/**
	 * Override of parent assign method
	 *
	 * @access public
     * @param string $name Variable to assign $val to
     * @param string / array $val Value to assign to $name
	 * @param boolean $magic_quotes
	 */
	public function assign ($name, $val = '', $magic_quotes = false) {

		if (!$this->_template_is_cached) {
			parent::assign($name, $val, $magic_quotes);
		}
	}

	/**
	 * Override of parent assign_file method
	 *
     * @access public
     * @param string $name Variable to assign $val to
     * @param string / array $val Values to assign to $name
	 */
	public function assign_file ($name, $val = '') {

		if (!$this->_template_is_cached) {
			parent::assign_file($name, $val);
		}
	}

	/**
	 * Override of parent parse method
	 *
     * @access public
     * @param string $bname Block name to parse
	 * @param int $cache_expiry Seconds to cache block for
	 */
	public function parse ($bname, $cache_expiry = 0) {

		if (!$this->_template_is_cached) {

			if (!$this->read_block_cache($bname, $cache_expiry)) {

				parent::parse($bname);

				$this->write_block_cache($bname, $cache_expiry);
			}
		}
	}

	/**
	 * Override of parent text method
	 *
     * @access public
     * @param string $bname Block name to return
     * @return string
	 */
	public function text ($bname = '') {

		$text = parent::text($bname);

		if (!$this->_template_is_cached && $this->cache_expiry > 0) {

			$this->write_template_cache();

		} elseif ($this->debug && $this->output_type == 'HTML') {

			$text_header = "<!-- CachingXTemplate debug:\n";

			if ($this->cache_expiry > 0) {

				$filename = $this->_get_filename();

				$file = $this->cache_dir . DIRECTORY_SEPARATOR . $filename . $this->cache_unique . $this->cache_ext;

				$text_header .= 'File: ' . $file . "\nExpires in: " . ($this->_cache_filemtime - $this->_cache_expiry) . " seconds -->\n";
			} else {
				$text_header .= "Template Cache (whole template) disabled -->\n";
			}

			$text = $text_header . $text;
		}

		return $text;
	}

	/**
	 * Read whole template cache file
	 *
	 * @access protected
	 */
	protected function read_template_cache () {

		$filename = $this->_get_filename();

		$file = $this->cache_dir . DIRECTORY_SEPARATOR . $filename . DIRECTORY_SEPARATOR . $this->cache_unique . $this->cache_ext;

		if ($this->cache_expiry > 0 && file_exists($file)) {

			$this->_cache_filemtime = filemtime($file);
			$this->_cache_expiry = time() - $this->cache_expiry;

			if ($this->_cache_filemtime >= $this->_cache_expiry) {
				if ($parsed_blocks = file_get_contents($file)) {
					$this->parsed_blocks = unserialize($parsed_blocks);
					$this->_template_is_cached = true;
				}
			} else {
				// Stale file
				if (is_writable($this->cache_dir) && is_writable($file)) {
					unlink($file);
				}
			}
		}
	}

	/**
	 * Write out whole template cache file
	 *
	 * @access protected
	 */
	protected function write_template_cache () {

		if ($this->cache_expiry > 0 && is_writable($this->cache_dir)) {

			$filename = $this->_get_filename();

			if (!file_exists($this->cache_dir . DIRECTORY_SEPARATOR . $filename)) {
				mkdir($this->cache_dir . DIRECTORY_SEPARATOR . $filename);
			}

			file_put_contents($this->cache_dir . DIRECTORY_SEPARATOR . $filename . DIRECTORY_SEPARATOR . $this->cache_unique . $this->cache_ext, serialize($this->parsed_blocks));
		}
	}

	/**
	 * Read block level cache file
	 *
	 * @access protected
	 * @param string $bname Block name to read from cache
	 * @param ing $cache_expiry Seconds to cache block for
	 * @return boolean
	 */
	protected function read_block_cache ($bname, $cache_expiry = 0) {

		$retval = false;

		$filename = $this->_get_filename();

		$file = $this->cache_dir . DIRECTORY_SEPARATOR . $filename . DIRECTORY_SEPARATOR . $bname . $this->cache_unique . $this->cache_ext;

		if ($cache_expiry > 0 && file_exists($file)) {

			$filemtime = filemtime($file);
			$cache_expiry = time() - $cache_expiry;

			if ($filemtime >= $cache_expiry) {
				if ($block = file_get_contents($file)) {
					$block = unserialize($block);
					if ($this->debug) {
						$block = "<!-- CachingXTemplate debug:\nFile: " . $file . "\nBlock: " . $bname . "\nExpires in: " . ($filemtime - $cache_expiry) . ' seconds -->' . "\n" . $block;
					}
					$this->parsed_blocks[$bname] = $block;
					$retval = true;
				}
			} else {
				// Stale file
				if (is_writable($this->cache_dir) && is_writable($file)) {
					unlink($file);
				}
			}
		}

		return $retval;
	}

	/**
	 * Write out block level cache file
	 *
	 * @access protected
	 * @param string $bname Block name to cache
	 * @param int $cache_expiry Seconds to cache block for
	 */
	protected function write_block_cache ($bname, $cache_expiry = 0) {

		if ($cache_expiry > 0 && is_writable($this->cache_dir)) {

			$filename = $this->_get_filename();

			if (!file_exists($this->cache_dir . DIRECTORY_SEPARATOR . $filename)) {
				mkdir($this->cache_dir . DIRECTORY_SEPARATOR . $filename);
			}

			file_put_contents($this->cache_dir . DIRECTORY_SEPARATOR . $filename . DIRECTORY_SEPARATOR . $bname . $this->cache_unique . $this->cache_ext, serialize($this->parsed_blocks[$bname]));
		}
	}

	/**
	 * Create the main part of the cache filename
	 *
	 * @access private
	 * @return string
	 */
	private function _get_filename () {

		$filename = $this->filename;
		if (!empty($this->tpldir)) {

			$filename = str_replace(DIRECTORY_SEPARATOR, '_', $this->tpldir . DIRECTORY_SEPARATOR) . $this->filename;
		}

		return $filename;
	}
}

?>